R/tribble.R

```diff
@@ -27,17 +27,80 @@
 #'   "b", 4:6
 #' )
 tribble <- function(...) {
+  data <- extract_frame_data_from_dots(...)
 
+  if (is.null(data$frame_rest)) {
+    out <- rep(list(logical()), length(data$frame_names))
+    names(out) <- data$frame_names
+    return(as_tibble(out))
+  }
+
+  frame_mat <- matrix(data$frame_rest, ncol = length(data$frame_names), byrow = TRUE)
+
+  frame_col <- lapply(seq_len(ncol(frame_mat)), function(i) {
+    col <- frame_mat[, i]
+    if (any(vapply(col, needs_list_col, logical(1L)))) {
+      col
+    } else {
+      unlist(col)
+    }
+  })
+
+  # Create a tbl_df and return it
+  names(frame_col) <- data$frame_names
+  as_tibble(frame_col)
+}
+
+frame_matrix <- function(...) {
+  data <- extract_frame_data_from_dots(...)
+
+  if (any(vapply(data$frame_rest, needs_list_col, logical(1)))) {
+    stopc("frame_matrix cannot have list columns")
+  }
+
+  frame_ncol <- length(data$frame_names)
+
+  if (is.null(data$frame_rest)) {
+    frame_mat <- matrix(logical(0), ncol = frame_ncol)
+  } else {
+    frame_mat <- matrix(unlist(data$frame_rest), ncol = frame_ncol, byrow = TRUE)
+  }
+
+  if (length(data$frame_names) != 0) {
+    colnames(frame_mat) <- data$frame_names
+  }
+
+  frame_mat
+}
+
+extract_frame_data_from_dots <- function(...) {
   dots <- list(...)
 
+  if (length(dots) == 0) {
+    return(list(frame_names = character(), frame_rest = NULL))
+  }
+
   # Extract the names.
+  frame_names <- extract_frame_names_from_dots(dots)
+
+  # Extract the data
+  if (length(frame_names) == length(dots)) {
```

Comment:
Is this special case really necessary? The code might look cleaner if you just did `frame_rest <- dots[-seq_along(frame_names)]` .